pkg/controller/deployment.go

```diff
@@ -604,3 +611,65 @@ func (dc *controller) updateMachineDeploymentFinalizers(ctx context.Context, mac
 		dc.updateMachineDeploymentFinalizers(ctx, machineDeployment, finalizers)
 	}
 }
+
+func (dc *controller) setMachinePriorityAnnotationAndUpdateTriggeredForDeletion(ctx context.Context, mcd *v1alpha1.MachineDeployment) error {
+	var toBeDeletedMachineNames, toBeSkippedMachineNames []string
+	toBeDeletedMachineNames = annotations.GetMachineNamesTriggeredForDeletion(mcd)
+	if len(toBeDeletedMachineNames) == 0 {
+		return nil
+	}
+	klog.V(3).Infof("MachineDeployment %q has #%d machine(s) marked for deletion, toBeDeletedMachineNames=%v", mcd.Name, len(toBeDeletedMachineNames), toBeDeletedMachineNames)
+	for _, machineName := range toBeDeletedMachineNames {
+		mc, err := dc.machineLister.Machines(dc.namespace).Get(machineName)
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("Machine %q is not found in MachineDeployment %q - skip setting MachinePriority=1 annotation", machineName, mcd.Name)
+			toBeSkippedMachineNames = append(toBeSkippedMachineNames, machineName)
+			continue
+		}
+		if machineutils.IsMachineFailedOrTerminating(mc) {
+			klog.V(4).Infof("Machine %q of MachineDeployment %q is in Failed/Terminating state - skip setting MachinePriority=1 annotation", machineName, mcd.Name)
+			toBeSkippedMachineNames = append(toBeSkippedMachineNames, machineName)
+			continue
+		}
+		if mc.Annotations[machineutils.MachinePriority] == "1" {
+			klog.V(4).Infof("Machine %q of MachineDeployment %q already marked with MachinePriority=1 annotation", machineName, mcd.Name)
+			continue
+		}
+		mcAdjust := mc.DeepCopy()
+		mcAdjust.Annotations[machineutils.MachinePriority] = "1"
+		_, err = dc.controlMachineClient.Machines(mcAdjust.Namespace).Update(ctx, mcAdjust, metav1.UpdateOptions{})
+		if err != nil {
+			klog.Errorf("Failed to set MachinePriority=1 annotation on Machine %q of MachineDeployment %q: %v", machineName, mcd.Name, err)
+			//TODO: Return error here or just log and continue for best-effort?
+			return err
+		}
+		klog.V(3).Infof("Machine %q of MachineDeployment %q marked with MachinePriority=1 annotation successfully", machineName, mcd.Name)
+	}
+
+	if len(toBeSkippedMachineNames) == 0 {
+		return nil
+	}
+
+	toBeDeletedMachineNames = sets.NewString(toBeDeletedMachineNames...).Delete(toBeSkippedMachineNames...).List()
+	triggerDeletionAnnotValue := annotations.CreateMachinesTriggeredForDeletionAnnotValue(toBeDeletedMachineNames)
+
+	if mcd.Annotations[machineutils.TriggerDeletionByMCM] == triggerDeletionAnnotValue {
```

Comment:
since you are using sets the order of machine names to be deleted will not be the same. Therefore this check (even though the machines to be deleted are the same) can be false because of the order. Is this check really required? You are already checking `len(toBeSkippedMachineNames) == 0` above.